### PR TITLE
E2E Test: Initial Adminisitrator Login Flow

### PR DESCRIPTION
**Pull request by nickytonline**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [x] E2E test for critical flow in Forem

## Description

Introduces our first end to end test to the Forem codebase. I went with the first thing that happens if you create a new Forem instance. Logging in as the initial administrator.

## Related Tickets & Documents

#382, #11841 

## QA Instructions, Screenshots, Recordings

To test:

- Pull down this branch on your local, for example, via the [GitHub CLI](https://github.com/cli/cli), `gh co 11840`
- Run the following from the command line to install the latest dependencies which include Cypress, `yarn`.
- Ensure your local server is running, i.e. `http://localhost:3000` by running `bin/startup` from the command line.
- Run `yarn e2e` on another command line.
- Cypress will open

image

- Click on the Run 1 Integrated Spec link

image

- The test runner will start

image

- The browser you have selected in the top right of the Cypress main window will run the tests successfully.

image

### UI accessibility concerns?

N/A

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

Charile Rose from Shitt's Creek saying "Let's do it"

### CODE REVIEW

File: cypress.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "baseUrl": "http://localhost:3000"
```
Comment (nickytonline) [inline cypress.json:2]: This is so we don't hard-code this in all our tests. To navigate to a page in Cypress, the `cy.visit(someUrl)` is used. If we write `cy.visist('/new')`, Cypress will no to navigate to `http://localhost:3000/new`.

File: cypress/fixtures/logins/initialAdmin.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (nickytonline) [inline cypress/fixtures/logins/initialAdmin.json:1]: Fixture data for the first admin user.

File: cypress/.eslintrc.js [hunk 3]
```diff
@@ -3,15 +3,14 @@
 module.exports = {
   parser: 'babel-eslint',
   extends: [
-    'plugin:cypress/recommended',
     'eslint:recommended',
     'plugin:import/errors',
     'plugin:import/warnings',
-    'preact',
+    'plugin:cypress/recommended',
```
Comment (nickytonline) [inline cypress/.eslintrc.js:9]: These changes will not be part of this PR once #11841 gets merged.

File: jest.config.js [hunk 4]
```diff
@@ -44,6 +44,7 @@ module.exports = {
     // Allows developers to add utility modules that jest won't run as test suites.
     '/__tests__/utilities/',
     '<rootDir>/app/javascript/storybook-static',
+    '<rootDir>/cypress',
```
Comment (nickytonline) [inline jest.config.js:47]: These changes will not be part of this PR once #11841 gets merged.

File: cypress/integration/adminLogin.spec.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import '@testing-library/cypress/add-commands';
+
+describe('Administrator login', () => {
+  it('should login the initial administrator user from the home page', () => {
+    cy.fixture('logins/initialAdmin.json').as('admin');
```
Comment (nickytonline) [inline cypress/integration/adminLogin.spec.js:5]: `.as(someString)` commands are aliases. This allows us to use the fixture via an aliased name. In the case of this fixture, we can now access it as `cy.get('@admin')`. Aliases are not just for fixture data though as you can see in the rest of the test below. For more info, please see the official Cypress documentation on [Variables and Aliases](https://docs.cypress.io/guides/core-concepts/variables-and-aliases.html#Aliases)

**Comment by rhymes** [comment]
I ran `yarn e2e` and got the following error:

```shell
Error: Webpack Compilation Error
./cypress/integration/adminLogin.spec.js
Module not found: Error: Can't resolve '@testing-library/cypress/add-commands' in '/Users/rhymes/Development/forem/cypress/integration'
resolve '@testing-library/cypress/add-commands' in '/Users/rhymes/Development/forem/cypress/integration'
  Parsed request is a module
  using description file: /Users/rhymes/Development/forem/package.json (relative path: ./cypress/integration)
    Field 'browser' doesn't contain a valid alias configuration
    Looked for and couldn't find the file at the following paths:
```

I also re-ran `yarn install --dev` and restarted the webserver

**Comment by nickytonline** [comment]
> I ran `yarn e2e` and got the following error:
> 
> ```shell
> Error: Webpack Compilation Error
> ./cypress/integration/adminLogin.spec.js
> Module not found: Error: Can't resolve '@testing-library/cypress/add-commands' in '/Users/rhymes/Development/forem/cypress/integration'
> resolve '@testing-library/cypress/add-commands' in '/Users/rhymes/Development/forem/cypress/integration'
>   Parsed request is a module
>   using description file: /Users/rhymes/Development/forem/package.json (relative path: ./cypress/integration)
>     Field 'browser' doesn't contain a valid alias configuration
>     Looked for and couldn't find the file at the following paths:
> ```
> 
> I also re-ran `yarn install --dev` and restarted the webserver

@rhymes, let me rerun stuff fresh locally to see if I may have omitted something.

**Comment by nickytonline** [comment]
Should be good now @rhymes. I forgot to commit `@testing-library/cypress` to dev dependencies.

**Comment by rhymes** [comment]
Thank you @nickytonline! I'll give it another whirl

File: cypress/plugins/index.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+/// <reference types="cypress" />
```
Comment (nickytonline) [inline cypress/plugins/index.js:1]: This files is automatically generated by Cypress. We don't need it yet, but I'm committing it in this PR because otherwise it will show up as a new file in git if folks run the tests locally.

File: cypress/support/commands.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+// ***********************************************
```
Comment (nickytonline) [inline cypress/support/commands.js:1]: This files is automatically generated by Cypress. We don't need it yet, but I'm committing it in this PR because otherwise it will show up as a new file in git if folks run the tests locally.

File: cypress/support/index.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// You can change the location of this file or turn off
+// automatically serving support files with the
+// 'supportFile' configuration option.
+//
+// You can read more here:
+// https://on.cypress.io/configuration
+// ***********************************************************
+
+// For Testing Library APIs https://github.com/testing-library/cypress-testing-library
+import '@testing-library/cypress/add-commands';
```
Comment (nickytonline) [inline cypress/support/index.js:17]: This files is automatically generated by Cypress. We don't need it yet, but I'm committing it in this PR because otherwise it will show up as a new file in git if folks run the tests locally.

Having said that, I added an import for Testing Library for Cypress so that folks don't need to do this explicitly in every test file.

File: app/views/shared/authentication/_email_login_form.html.erb [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<%= form_for(User.new, as: :user, url: session_path(:user)) do |f| %>
+<%= form_for(User.new, as: :user, url: session_path(:user), data: { testid: "login-form" }) do |f| %>
```
Comment (nickytonline) [inline app/views/shared/authentication/_email_login_form.html.erb:1]: Going with `data-testid`. This is common practice and we do this already in our other tests. For more information, see [Making your UI tests resilient to change](https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change).

**Comment by nickytonline** [comment]
The code climate issue is complaining about an eslint rule that does not exist. Looks like we may need to add some eslint rules for Cypress on code climate. It doesn't block this at all though.

**Review by rhymes — APPROVED** [review]
LGTM!

**Review by fdoxyz — APPROVED** [review]
This is pretty cool to run locally @nickytonline !

**Question non-blocking:** I realized I don't use `admin@forem.local` in my environment so I got sent directly to the onboarding, but I guess that's on me 😅 This shouldn't impact CI right? In case this comes up at some point, we'll probably want to edit `db/seeds.rb` ([here](https://github.com/forem/forem/blob/master/db/seeds.rb#L167-L184)) to make sure that user doesn't have to go through onboarding

Screen Shot 2020-12-11 at 07 51 02

**Comment by nickytonline** [comment]
> This is pretty cool to run locally @nickytonline !
> 
> **Question non-blocking:** I realized I don't use `admin@forem.local` in my environment so I got sent directly to the onboarding, but I guess that's on me 😅 This shouldn't impact CI right? In case this comes up at some point, we'll probably want to edit `db/seeds.rb` ([here](https://github.com/forem/forem/blob/master/db/seeds.rb#L167-L184)) to make sure that user doesn't have to go through onboarding
> 

@fdoxyz, yeah that won't/shouldn't affect CI. Also, this flow is just testing the login. We'll definitely test out the onboarding flow in another e2e test.

**Comment by nickytonline** [comment]
It doesn't affect the test, but I will update the test to include the onboarding in the URL

image
